src/core/regimes.rkt

```diff
@@ -173,64 +173,80 @@
   (define (add-splitpoint sp-prev)
+    (struct idk (cidx pidx) #:transparent)
     ;; If there's not enough room to add another splitpoint, just pass the sp-prev along.
-    (for/vector #:length num-points ([point-idx (in-naturals)] [point-entry (in-vector sp-prev)])
+    (define result (make-vector (+ num-points)))
+    (for ([point-idx (in-naturals)] [point-entry (in-vector sp-prev)])
       ;; We take the CSE corresponding to the best choice of previous split point.
       ;; The default, not making a new split-point, gets a bonus of min-weight
       (let ([acost (- (cse-cost point-entry) min-weight)] [aest point-entry])
-        (for ([prev-split-idx (in-range 0 point-idx)] [prev-entry (in-vector sp-prev)]
-              #:when (can-split? (si-pidx (car (cse-indices prev-entry)))))
+        (for ([prev-split-idx (in-range 0 point-idx)] 
+              [prev-entry (in-vector sp-prev)]
+              #:when (when-cond prev-entry))
           ;; For each previous split point, we need the best candidate to fill the new regime
           (let ([best #f] [bcost #f])
-            (for ([cidx (in-naturals)] [psum (in-list psums)])
+            (for ([cidx (in-naturals)] [psum (in-vector vec-psums)])
               (let ([cost (- (vector-ref psum point-idx)
                              (vector-ref psum prev-split-idx))])
                 (when (or (not best) (< cost bcost))
                   (set! bcost cost)
                   (set! best cidx))))
             (when (and (< (+ (cse-cost prev-entry) bcost) acost))
               (set! acost (+ (cse-cost prev-entry) bcost))
-              (set! aest (cse acost (cons (si best (+ point-idx 1))
+              (define idk (si best (+ point-idx 1)))
+              (cond ((struct? idk) (set! idk (vector idk))))
+              (set! aest (cse acost (vector-append idk 
                                           (cse-indices prev-entry)))))))
-        aest)))
+        (vector-set! result point-idx aest)
+        aest))
+    result)
 
   ;; We get the initial set of cse's by, at every point-index,
   ;; accumulating the candidates that are the best we can do
   ;; by using only one candidate to the left of that point.
-  (define initial
-    (for/vector #:length num-points ([point-idx (in-range num-points)])
-      (argmin cse-cost
+  (define (initial idk)
+    (define output (make-vector (+ num-points)))
+    (for ([point-idx (in-range num-points)])
+      (define o (vector-argmin cse-cost
               ;; Consider all the candidates we could put in this region
-              (map (λ (cand-idx cand-psums)
+              (vector-map (λ (cand-idx cand-psums)
                       (let ([cost (vector-ref cand-psums point-idx)])
-                        (cse cost (list (si cand-idx (+ point-idx 1))))))
-                   (range num-candidates)
-                   psums))))
+                        (cse cost (vector (si cand-idx (+ point-idx 1))))))
+                   (list->vector (range num-candidates))
```

Comment:
See this constructs a list (allocates) and then converts it to a vector (copies). Rewrite this to `for/vector` (with an explicit size) and you skip that.